src/Init/Data/List/MinMax.lean

```diff
@@ -155,6 +155,49 @@ theorem foldl_min [Min α] [Std.IdempotentOp (min : α → α → α)] [Std.Asso
     {l : List α} {a : α} : l.foldl (init := a) min = min a (l.min?.getD a) := by
   cases l <;> simp [min?, foldl_assoc, Std.IdempotentOp.idempotent]
 
+/-! ### min -/
+
+theorem min?_eq_some_min [Min α] : {l : List α} → (hl : l ≠ []) →
+    l.min? = some (l.min hl)
+  | a::as, _ => by simp [List.min, List.min?_cons']
+
+theorem min_eq_head {α : Type u} [Min α] {l : List α} (hl : l ≠ [])
+    (h : l.Pairwise (fun a b => min a b = a)) : l.min hl = l.head hl := by
+  apply Option.some.inj
+  rw [← min?_eq_some_min, ← head?_eq_some_head]
+  exact min?_eq_head? h
+
+theorem min_mem [Min α] [MinEqOr α] {l : List α} (hl : l ≠ []) : l.min hl ∈ l :=
+  min?_mem (min?_eq_some_min hl)
+
+protected theorem le_min_iff [Min α] [LE α] [LawfulOrderInf α]
+    {l : List α} (hl : l ≠ []) : ∀ {x}, x ≤ l.min hl ↔ ∀ b, b ∈ l → x ≤ b :=
+  le_min?_iff (min?_eq_some_min hl)
+
+theorem min_eq_iff  [Min α] [LE α] {l : List α} [IsLinearOrder α] [LawfulOrderMin α] (hl : l ≠ []) :
+    l.min hl = a ↔ a ∈ l ∧ ∀ b, b ∈ l → a ≤ b := by
+  simpa [min?_eq_some_min hl] using (min?_eq_some_iff (xs := l))
+
+theorem min_eq_min_attach [Min α] [MinEqOr α] {l : List α} (hl : l ≠ []) :
+    l.min hl = Subtype.val (l.attach.min (List.attach_ne_nil_iff.mpr hl)) := by
+  simpa [min?_eq_some_min hl, min?_eq_some_min (List.attach_ne_nil_iff.mpr hl)]
+    using (min?_eq_min?_attach (xs := l))
```

Comment:
I'm not sure what this theorem is for.

I wonder if the version with `l.attach.min ...` on the left-hand side (and no `Subtype.val`) might be more useful (even as a simp lemma).